x/incentives/keeper/distribute.go

```diff
@@ -9,11 +9,127 @@ import (
+// Any issues can be fixed with a major upgrade without any loss of funds.
+// - group gauge is inactive - similar reason, don't want to fail all distributions due to non-fatal issue.
+//
+// ASSUMPTIONS:
+// - group cannot outlive the gauges associated with it.
+// CONTRACT:
+// - every group in the input is active. If inactive group is passed in, it will be skipped.
+func (k Keeper) AllocateAcrossGauges(ctx sdk.Context, activeGroups []types.GroupGauge) error {
+	for _, group := range activeGroups {
+		err := k.syncGroupGaugeWeights(ctx, group)
+		if err != nil {
+			ctx.Logger().Error("error syncing group gauge weights, skipping", "group gauge id", group.GroupGaugeId, "error", err.Error())
+			continue
+		}
+
+		// Get the groupGauge corresponding to the group.
+		groupGauge, err := k.GetGaugeByID(ctx, group.GroupGaugeId)
+		if err != nil {
+			return err
+		}
+
+		// If upcoming, skip.
+		if !groupGauge.IsActiveGauge(ctx.BlockTime()) {
+			continue
+		}
+
+		// Get amount to distribute in coins (based on perpetual or non perpetual group gauge)
+		coinsToDistribute := groupGauge.Coins.Sub(groupGauge.DistributedCoins)
+		if !groupGauge.IsPerpetual {
+			remainingEpochs := int64(groupGauge.NumEpochsPaidOver - groupGauge.FilledEpochs)
+
+			// Divide each coin by remainingEpochs
+			osmoutils.QuoRawMut(coinsToDistribute, remainingEpochs)
+		}
+
+		// Exit early if nothing to distribute.
+		if coinsToDistribute.IsZero() {
+			// Update the group gauge despite zero coins distributed to ensure the filled epochs are updated.
+			if err = k.updateGaugePostDistribute(ctx, *groupGauge, coinsToDistribute); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// We track this for distributing all the remaining amount in the last
+		// gauge without leaving truncation dust.
+		amountDistributed := sdk.NewCoins()
+
+		// Define variables for brevity
+		totalGroupWeight := group.InternalGaugeInfo.TotalWeight
+		gaugeCount := len(group.InternalGaugeInfo.GaugeRecords)
+
+		// Iterate over underlying gauge records in the group.
+		for gaugeIndex, distrRecord := range group.InternalGaugeInfo.GaugeRecords {
+			// Between 0 and 1. to determine the pro-rata share of the total amount to distribute
+			gaugeDistributionRatio := distrRecord.CurrentWeight.ToLegacyDec().Quo(totalGroupWeight.ToLegacyDec())
+
+			// Loop through `coinsToDistribute` and get the amount to distribute to the current gauge
+			// based on the distribution ratio.
+			currentGaugeCoins := osmoutils.MulDec(coinsToDistribute, gaugeDistributionRatio)
+
+			moduleAccount := k.ak.GetModuleAddress(types.ModuleName)
+
+			// For the last gauge, distribute all remaining amounts.
+			// Special case the last gauge to avoid leaving truncation dust in the group gauge
+			// and consume the amounts in-full.
+			if gaugeIndex == gaugeCount-1 {
+				// TODO: module account sends funds to itself. Should we expose API that does not require this?
```

Comment:
I recall Sishir started a function called `AddToGaugeRewardsFromGauge` for this but if this works then I'm not sure why this would be necessary. Perhaps it would be sufficient to simply add a check to `AddToGaugeRewards` that skips if sender and receiver are the same?

Reply:
Nvm I see you addressed this nicely in #6445